### PR TITLE
gRPC Python test infrastructure

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
Code has been moved around but I don't think anything's really changed in a meaningful way since [the last draft that got a boatload of comments](https://github.com/nathanielmanistaatgoogle/grpc/commit/d45216fa1d37519f2d0798131831efdc4a8ec016).

@griffithjames, please also review.

### CODE REVIEW

File: src/python/grpcio_tests/tests/testing/proto/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2017 gRPC authors.
```
Comment (kpayson64) [inline src/python/grpcio_tests/tests/testing/proto/__init__.py]: Is this file needed?

File: src/python/grpcio_tests/tests/testing/_client_test.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+from tests.testing.proto import services_pb2
+
+
+class ClientTest(unittest.TestCase):
+
+    def setUp(self):
+        # In this test the client-side application under test executes in
+        # a separate thread while we retain use of the test thread to "play
+        # server".
+        self._client_execution_thread_pool = logging_pool.pool(1)
```
Comment (kpayson64) [inline src/python/grpcio_tests/tests/testing/_client_test.py]: It seems like using the testing infrastructure to test any blocking RPCs will require something like this.

It might be nice if I could ```take_next_rpc``` in advance or something like that, so that I could run the client app in the main thread.  This would probably make debugging easier, especially in the case of apps with unhanded exceptions.

I realize this would be a large structural change, so I'm not suggesting we implement this, but I wanted to mention it.

File: src/python/grpcio_testing/grpc_testing/_channel/_channel.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import grpc_testing
+from grpc_testing._channel import _channel_rpc
+from grpc_testing._channel import _multi_callable
+
+
+# All per-call credentials parameters are unused by this test infrastructure.
```
Comment (mehrdada) [inline src/python/grpcio_testing/grpc_testing/_channel/_channel.py]: credentials parameters? (I think this may be copied and pasted from somewhere else)

File: src/python/grpcio_testing/grpc_testing/_common.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+def fuss_with_metadata(metadata):
+    if metadata is None:
+        return FUSSED_EMPTY_METADATA
+    else:
+        return _fuss(tuple(metadata))
+
+
+class ChannelRpcRead(
+        collections.namedtuple(
```
Comment (mehrdada) [inline src/python/grpcio_testing/grpc_testing/_common.py]: Is this the appropriate indentation convention for this construct? (genuine question)

File: src/python/grpcio_testing/grpc_testing/__init__.py [hunk 5]
```diff
@@ -117,3 +384,19 @@ def strict_fake_time(now):
     """
     from grpc_testing import _time
     return _time.StrictFakeTime(now)
+
+
+def channel(service_descriptors, time):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/__init__.py]: My thoughts are to punt on it for now and maybe come back to it later. It's nonbreaking-to-users to do this now, find out it was a mistake, and add a default later, but the reverse is not the case.

File: src/python/grpcio_testing/grpc_testing/_channel/_channel.py (hunk 3 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_channel/_channel.py]: Yep, corrected to "All serializer and deserializer parameters are not (yet) used by this test infrastructure.". Thanks for the catch.

File: src/python/grpcio_testing/grpc_testing/_channel/_channel.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+# pylint: disable=unused-argument
+class TestingChannel(grpc_testing.Channel):
+
+    def __init__(self, time, state):
+        self._time = time
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_channel/_channel.py]: Yep. Leaving for now as I do think this will be used later.

File: src/python/grpcio_testing/grpc_testing/_channel/_rpc_state.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    def termination(self):
+        with self._condition:
+            while True:
+                if self._code is None:
+                    self._condition.wait()
+                else:
+                    return self._trailing_metadata, self._code, self._details
+
+    def cancel(self, code, details):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_channel/_rpc_state.py]: Maybe I should have called this "abort" - it's more akin to "cancel for the given reason", kind of like [`grpc_call_cancel_with_status`](https://github.com/grpc/grpc/blob/4042e4c4d4b36891a2c220c8d5b69b9ad73a77bd/include/grpc/grpc.h#L294-L303) in the core API.

File: src/python/grpcio_testing/grpc_testing/_common.py (hunk 4 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_common.py]: No; fixed.

PEP8 and I have some differences, I guess, and `yapf` maybe has some holes in its coverage?

File: src/python/grpcio_tests/tests/testing/_client_test.py (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/testing/_client_test.py]: Agreed that there is something to be done here and that it is neither easy nor small. 😄

File: src/python/grpcio_tests/tests/testing/proto/__init__.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/testing/proto/__init__.py]: Maybe not? I was just following the existing pattern in tests/protoc_plugin/protos.

But when I remove the file, the test fails, so [let's keep it for now and if it can be cleaned up we'll clean it up later](https://github.com/grpc/grpc/issues/11993).

File: src/python/grpcio_testing/grpc_testing/_common.py (hunk 4 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_common.py]: Welp; it turns out that this is the way `pylint` wants it. Restored to eight spaces.

File: src/python/grpcio_testing/grpc_testing/_channel/_rpc_state.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+import grpc
+from grpc_testing import _common
+
+
+class State(_common.ChannelRpcHandler):
+
+    def __init__(self, invocation_metadata, requests, requests_closed):
+        self._condition = threading.Condition()
+        self._invocation_metadata = invocation_metadata
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_testing/grpc_testing/_channel/_rpc_state.py]: It could, but there's no need for it to be. Each of the three [metadata objects](https://grpc.io/grpc/python/grpc.html#glossary) in an RPC is treated as a (probably immutable) [sequence](https://docs.python.org/2/library/stdtypes.html#sequence-types-str-unicode-list-tuple-bytearray-buffer-xrange). Metadata is accepted whole from the application on one side, delivered whole to the application on the other side, and mostly not molested by the gRPC code in between.

In the particular circumstance of this instance-scope field: non-`None` indicating metadata accepted from the system under test and not yet delivered to the testing system and `None` indicating metadata having already been delivered to the testing system are exactly the semantics we want. We want semantics simpler than one usually gets with a mutable list-valued field.

**Comment by nathanielmanistaatgoogle** [comment]
Basic Tests Mac failure: [12077](https://github.com/grpc/grpc/issues/12077).

Merging.
